### PR TITLE
[Feature] exercise view

### PR DESCRIPTION
**Pull request by at-chinhpham**
exercise view basic: 3, 7, init project

### CODE REVIEW

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    @IBOutlet private weak var progressBar: UIView!
+    // MARK: - Properties
+    
+    // MARK: - Life Cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.navigationController?.setNavigationBarHidden(true, animated: false)
+        valueChanged.text = "\(Int((ballSlider.frame.minY / 310) * 100))%"
+        progressBar.layer.borderColor = UIColor.clear.cgColor
+        progressBar.layer.cornerRadius = 8
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift]: Define number

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        progressBar.layer.borderWidth = 1
+        
+        //
+        progress.layer.cornerRadius = 8
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift]: make a function for same code

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    // MARK: - Private, Public
+    
+    // MARK: - IBActions
+    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
+        if let touches = touches.first {
+            let position = touches.location(in: self.view)
+            if position.y >= 175 && position.y <= 486 {
+                
+                // Progress
+                progress.frame = CGRect(x: 4, y: 4, width: 21, height: position.y - 165)
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift]: define

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                
+                ballSlider.frame = CGRect(x: 100, y: position.y - 175, width: 60, height: 60)
+                valueChanged.text = "\(Int((ballSlider.frame.minY / 310) * 100))%"
+                print(ballSlider.frame.minY)
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift]: remove print

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    @IBOutlet private weak var valueChanged: UILabel!
+    @IBOutlet private weak var progress: UIView!
+    @IBOutlet weak var bgMain: UIView!
+    @IBOutlet private weak var progressBar: UIView!
+    // MARK: - Properties
+    
+    // MARK: - Life Cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.navigationController?.setNavigationBarHidden(true, animated: false)
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift]: remove self.

File: Chinh Pham N./MyLearning/MyLearning/Define.swift [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    
+    struct Login {
+    }
+}
+
+// Home
+extension Define.String.Home {
+    static let title = "ASIAN TECH"
+    static let buttonExercise3 = "OPEN EXERCISE 3: VIEW BASIC"
+        static let buttonExercise7 = "OPEN EXERCISE 7: SLIDER"
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Define.swift]: clear space

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    func clearForm() {
+        usernameTextfield.text = ""
+        passwordTextField.text = ""
+    }
+    
+    // Validate username & password
+    private func validateLogin(uname: String, password: String) -> Bool {
+        var isSuccess = true
+        let dummyData = Define.String.Login.DummyData.self
+        if uname.rangeOfCharacter(from: CharacterSet.whitespaces) != nil {
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift]: change to enum

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+            alertMessage(title: Define.String.Login.Message.alertErrorsTitle, msg: Define.String.Login.Message.usernameDoesNotContainSpace)
+            isSuccess = false
+        } else if uname.count < 6 || uname.count > 32 {
+            alertMessage(title: Define.String.Login.Message.alertErrorsTitle, msg: Define.String.Login.Message.usernameWrongCharacter)
+            isSuccess = false
+        } else if uname.lowercased() != dummyData.username || password != dummyData.password {
+            alertMessage(title: Define.String.Login.Message.alertErrorsTitle, msg: Define.String.Login.Message.wrongUsernameOrPassword)
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift]: validation not call alert

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIKeyboardWillShow, object: nil)
+        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIKeyboardWillHide, object: nil)
+    }
+    
+    // Button login
+    @objc private func btnDidLoginClick() {
+        guard let username = usernameTextfield.text, let password = passwordTextField.text else { return }
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift]: bring to validation

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+    
+    // MARK: - Private
+    @objc private func keyboardDidShow(notification: Notification) {
+        let info: NSDictionary = notification.userInfo! as NSDictionary
+        let keyboardSize = (info[UIKeyboardFrameBeginUserInfoKey] as! NSValue).cgRectValue
+        let keyboardY = self.view.frame.size.height - keyboardSize.height
+        
+        let edittingTextFieldY: CGFloat! = self.activeButtonScale?.frame.origin.y
+        
+        if self.view.frame.origin.y >= 0 {
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift]: remove self

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+            if edittingTextFieldY > keyboardY - 60 {
+                UIView.animate(withDuration: 0.25, delay: 0.0,
+                               options: UIViewAnimationOptions.curveEaseIn, animations: {
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift]: remove UIViewAnimationOptions

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+        
+        // Notification center
+        let center: NotificationCenter = NotificationCenter.default
+        center.addObserver(self, selector: #selector(keyboardDidShow(notification:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
+        center.addObserver(self, selector: #selector(keyboardWillHide(notification:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
+    }
+    
+    // MARK: - Private
+    @objc private func keyboardDidShow(notification: Notification) {
+        let info: NSDictionary = notification.userInfo! as NSDictionary
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Login/LoginViewController.swift]: use guard let or if let 
dont let "!" invisible

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Home/HomeViewController.swift [hunk 13]
```diff
@@ -1,35 +1,147 @@
...
+        label.font = UIFont.systemFont(ofSize: Config.authorTextSize)
+        label.text = Define.String.author
+        label.textColor = UIColor.white
+        return label
+    }
+    
+    // Background image
+    private var backgroundImage: UIImageView {
+        let imageView = UIImageView(frame: Config.backgroundFrame)
+        imageView.backgroundColor = UIColor(patternImage: UIImage(named: "bgHome")!)
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Home/HomeViewController.swift]: ???

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+//  Created by Chinh Pham N. on 4/4/18.
+//  Copyright © 2018 Asiantech. All rights reserved.
+//
+
+import UIKit
+
+final class SliderViewController: UIViewController {
+
+    // MARK: - IBOutlets
+    @IBOutlet private weak var ballSlider: UIView!
```
Comment (at-quangdang) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Slider/SliderViewController.swift]: Change name 
biến outlet = name + property
Ví dụ:
sliderView
valueChangedLabel

File: Chinh Pham N./MyLearning/MyLearning/Controllers/Detail/DetailViewController.swift [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        super.viewDidLoad()
+        guard let username = username else { return }
+        welcomeLabel.text = "Welcome \(username)"
+    }
+    
+    // MARK: - Private, Public
+    
+    // MARK: - IBActions
+    @IBAction func logoutButton(_ sender: Any) {
+        self.navigationController?.popViewController(animated: true)
```
Comment (at-tamnguyen3) [inline Chinh Pham N./MyLearning/MyLearning/Controllers/Detail/DetailViewController.swift]: Omit `self`.

**Pull request merged**